Fix break in DirectoryServices SearchResultCollection

IDisposable opened the pull request:
Added enumeration handling when the internal ArrayList has already been loaded but an enumeration follows. This fixes the issue with 9.x Linq statements against the collection caused by the access of the .Count

Fixes #113265

I am not entirely sure the best way to test this, as we never had tests for this class until now so I created a new DirectorySearcherTests.cs (conditionally run when LDAP is available) using the setup shown in the bug report.

IDisposable committed:
Fix break in DirectoryServices SearchResultCollection

Added enumeration handling when the internal ArrayList has already been loaded but an enumeration follows. This fixes the issue with 9.x Linq statements against the collection caused by the access of the .Count

Fixes #113265

IDisposable commented:
I think the code is the simplest way to fix this issue without making tons more changes. I'm not sure my test class is useful/valid because I'm not on a LDAP domain-joined machine so it simply bypasses for me in local runs.

:pointup: @BillArmstrong @stephentoub @tarekgh

tarekgh commented on src/libraries/System.DirectoryServices/src/System/DirectoryServices/SearchResultCollection.cs:
```diff
@@ -214,6 +214,49 @@ void ICollection.CopyTo(Array array, int index)
+        /// <devdoc>
+        /// Supports a simple type-specific wrapper for the underlying cached list
+        /// </devdoc>
+        private sealed class AlreadyReadResultsEnumerator : IEnumerator
```
>class [](http://example.com/codeflow?start=23&length=5)

can this be a struct?

IDisposable replied:
Good point, I will adjust after the Blues game.

IDisposable replied:
Switched to `struct`

tarekgh reviewed (approved):
LGTM, thanks @IDisposable 

@stephentoub @steveharter could you please have a quick look? Note, this is a candidate for 9.0 service porting .

IDisposable committed:
PR feedback

Switched enumerator to struct
Improved testing actual scenario of the bug.

IDisposable commented:
Push PR review updates and improved the unit tests. Can anyone do the `/azp` magic to trigger the tests for `IsLdapConfigurationExist` since I can't run those local, nor does the normal build run them?

IDisposable commented on src/libraries/System.DirectoryServices/src/System/DirectoryServices/SearchResultCollection.cs:
```diff
@@ -214,6 +214,49 @@ void ICollection.CopyTo(Array array, int index)
             InnerList.CopyTo(array, index);
         }
 
+        /// <devdoc>
```
I just took the existing enumerator's framework and dropped it back to standard enumeration. I can add `inheritdocs` if desired.

IDisposable commented on src/libraries/System.DirectoryServices/tests/System.DirectoryServices.Tests.csproj:
```diff
@@ -9,6 +9,7 @@
     <Compile Include="System\DirectoryServices\ActiveDirectory\DomainControllerTests.cs" />
     <Compile Include="System\DirectoryServices\DirectoryEntryTests.cs" />
     <Compile Include="System\DirectoryServices\DirectoryServicesPermissionTests.cs" />
+    <Compile Include="System\DirectoryServices\DirectorySearcherTests.cs" />
```
Oh, sorry... I just added it, will resort.

IDisposable commented on src/libraries/System.DirectoryServices/src/System/DirectoryServices/SearchResultCollection.cs:
```diff
@@ -214,6 +214,49 @@ void ICollection.CopyTo(Array array, int index)
+        /// <devdoc>
+        /// Supports a simple type-specific wrapper for the underlying cached list
+        /// </devdoc>
+        private struct AlreadyReadResultsEnumerator : IEnumerator
```
We certainly could, except this enumerator matches the existing lazy one in that the `Current` method is declared to return a `SearchResult` (e.g. correct type information). If I just used `ArrayList.GetEnumerator`, then `Current` would be returning `object`, which is inconsistent between the two (lazy/eager) enumerators.

Please let me know if you want that change made.

IDisposable commented on src/libraries/System.DirectoryServices/tests/System/DirectoryServices/DirectorySearcherTests.cs:
```diff
@@ -0,0 +1,76 @@
+        private const int ADS_SYSTEMFLAG_CR_NTDS_DOMAIN = 0x2;
+
+        [ConditionalFact(nameof(IsLdapConfigurationExist))]
+        public void DirectorySearch_IteratesCorrectly_SimpleEnumeration()
```
There was never any tests against this class as far as I could find, so no :)

The two tests come from from @stephentoub's minimal reproduction [in this comment](https://github.com/dotnet/runtime/issues/113265#issuecomment-2714683962)

IDisposable replied:
Sorry, that might have been unclear... the first test would have always succeeded (and I put this test in to ensure I didn't break anything). The second test would have failed without the changes in this PR.

IDisposable commented on src/libraries/System.DirectoryServices/tests/System/DirectoryServices/DirectorySearcherTests.cs:
```diff
@@ -0,0 +1,76 @@
+        private static SearchResultCollection GetDomains()
+        {
+            using var entry = new DirectoryEntry("LDAP://rootDSE");
+            var namingContext = entry.Properties["configurationNamingContext"][0]!.ToString();
```
This was from the original bug report, would it be better to mimic that code or write it "better"?

steveharter replied:
We should follow the coding guidelines, especially for obvious cases for readability\understandability, although for some things like using underscores in test names we don't necessarily follow.

IDisposable replied:
Done

IDisposable committed:
PR feedback

Switched the enumerator doc comments to XML docs.
Resorted the test class in the csproj
Strong type the SearchResultCollection in the test class.

IDisposable commented on src/libraries/System.DirectoryServices/src/System/DirectoryServices/SearchResultCollection.cs:
```diff
@@ -47,15 +47,17 @@ private ArrayList InnerList
-                        _innerList.Add(enumerator.Current);
+                        eagerList.Add(enumerator.Current);
+
+                    _innerList = eagerList;
```
The change to build the inner list in it's entirety _and then_ set the member is to ensure that if an exception in thrown while populating this inner list we do NOT memoize the incomplete list because then we would (if enumerated again) not throw again and would instead silently only yield the entries we previously gathered. Of course anyone attempting _re_enumeration of a list that yielded an error the first time is probably wrong, but why add to the problem?

IDisposable committed:
PR feedback

Made the unit test helper use actual variable types (not var)

steveharter commented:
PTAL @BRDPM @grubioe @jay98014

steveharter commented:
FYI link to coding guidelines: https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md which do discuss the use of "var".

stephentoub commented on src/libraries/System.DirectoryServices/src/System/DirectoryServices/SearchResultCollection.cs:
```diff
@@ -214,6 +214,51 @@ void ICollection.CopyTo(Array array, int index)
         }
 
+        /// <summary>Provides an enumerator implementation for the array of <see cref="SearchResult"/>.</summary>
+        private struct AlreadyReadResultsEnumerator : IEnumerator
```
Why struct?

tarekgh replied:
This was my suggestion; I thought it may save some allocations in some scenarios?

stephentoub replied:
It looks like the only place it's used is here:
https://github.com/dotnet/runtime/pull/113775/files#diff-c9880e3568259df335f4e47504988814365bc6213087cd341b8939595f1fd988R194
in which case it's always immediately cast to an IEnumerator, boxing it.

tarekgh replied:
We can change it back to class if it makes a difference.

IDisposable replied:
Happy to swap it back, I did look for some other uses when it was suggested.

tarekgh replied:
go ahead and change it back.

IDisposable replied:
PR #113927 submitted.
